### PR TITLE
chore(TS): migrate brushes

### PR DESCRIPTION
**Pull request by ShaMan123**
I will migrate eraser on a separate PR
It will wait...

### CODE REVIEW

File: src/brushes/pencil_brush.class.ts [hunk 1]
```diff
@@ -1,314 +1,295 @@
...
+    this.oldEnd = undefined;
+    this._finalizeAndAddPath();
+    return false;
+  }
 
-    /**
-     * @private
-     * @param {Object} pointer Actual mouse position related to the canvas.
-     */
-    _captureDrawingPath: function(pointer) {
```
Comment (ShaMan123) [inline src/brushes/pencil_brush.class.ts:167]: removed
this was a redundant wrapper around `_addPoint`

**Review by ShaMan123 — COMMENTED** [review]
See comments for major changes
I propose we add an options object in the constructor passing initial state. I have been wanting this for a while

**Comment by asturur** [comment]
I have no issue with adding initial state to a brush ( even if it can probably be done outisde the constructor ).
The point is that the brush interface as it is now does not make sense, they are way similar to objects and have this issue of starting and ending with mouse down/up.
I want to change that and i m not sure how things will translate, but in general we can assume is safe to think brushes can have options when created.

File: src/brushes/base_brush.class.ts [hunk 2]
```diff
@@ -1,148 +1,150 @@
...
+    if (!this.shadow || !this.canvas) {
+      return;
     }
-  });
-})(typeof exports !== 'undefined' ? exports : window);
+
+    const canvas = this.canvas,
+      shadow = this.shadow,
+      ctx = canvas.contextTop,
+      zoom = canvas.getZoom() * canvas.getRetinaScaling();
```
[the remark this answers is not in the document]
Comment (asturur) [reply src/brushes/base_brush.class.ts:117]: yes but it should be 100% equivalent

**Comment by ShaMan123** [comment]
I don't mind waiting and returning to it later.
I want to hear more of your thoughts regarding brushes

File: src/brushes/base_brush.class.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ShaMan123) [reply src/brushes/base_brush.class.ts:117]: It should

**Comment by ShaMan123** [comment]
let's merge this. enough changes for one iteration

**Comment by asturur** [comment]
yes this will get merge with no additional changes today.
Regrding brushes, i would like a brush to expose an even handler for every mouse or touch event.
down, click, up, dblclick and so on.
Those events will get executed as today before standard canvas logic, and will by default block the event from running the canvas logic afterward, but is optional to the dev to do so.
The brush will be in control to terminate its own lifecycle on one of those events.

This should let current brushes work as they are working today, but would also allow brushes to work on clicks and collect points like to build a polygon and end when the polygon is closed or end on dblclick.

It doesn't change much but offer so much more flexibility.

**Review by asturur — APPROVED** [review]

**Comment by ShaMan123** [comment]
sounds great
and even not block events so things can run together creating complex UX.
This can be leveraged to many things like a custom pointer etc.

**Pull request merged**